src/l3.sc

```diff
@@ -0,0 +1,190 @@
+import scala.annotation.tailrec
+//Lista 3 Maciej Kopiński
+
+//useful
+def reverse[A](list : List[A]) : List[A] = {
+  @tailrec
+  def inner(acc : List[A], list : List[A]) : List[A] = {
+    list match {
+      case Nil => acc
+      case (h :: t) => {
+        inner(h :: acc, t)
+      }
+    }
+  }
+  inner(Nil, list)
+}
+
+def reverseTwo[A](list1: List[A], list2: List[A]):(List[A], List[A])={
+  @tailrec
+  def inner(accum1: List[A], accum2: List[A], xs: List[A], ys: List[A]):(List[A], List[A])={
+    (xs, ys) match{
+      case (Nil, Nil)=> (accum1, accum2)
+      case (Nil, h::t)=> inner(accum1, h::accum2, xs, t)
+      case (h::t, Nil)=> inner(h::accum1, accum2, t, ys)
+      case (h1::t1, h2::t2)=> inner(h1::accum1, h2::accum2, t1, t2)
+    }
+  }
+  inner(Nil, Nil, list1, list2)
+}
+
+//zlozonosc czasowa i pamieciowa liniowa
+def contains(main: String, substring: String): Boolean = {
+  (main, substring) match {
+    case (_, "")=> true
+    case ("", _)=> false
+    case (_, _)=> {
+      def ifPrefix(str1: String, str2: String):Boolean={
+        (str1, str2) match {
+          case ("", _)=> false
+          case (_, "")=> true
+          case (_, _)=> if(str1.head==str2.head) ifPrefix(str1.tail, str2.tail) else false
+        }
+      }
+      if(ifPrefix(main, substring)) true
+      else (ifPrefix(main.tail, substring.tail))
+    }
+  }
+}
+
+//zlozonosc kwadratowa
+def containsAny(main: String, substrings: List[String]): Boolean = {
+  (main, substrings) match {
+    case ("", _)=> false
+    case (_, Nil)=> false
+    case (_, h::t)=> {
+      if(contains(main, h)) true
+      else containsAny(main, t)
+    }
+  }
+}
+
+
+//ZADANIE 1
+
+//zlozonosc czasowa liniowa wzgledem listy wejsciowej
+//zlozonosc pamieciowa stala
+def divide(xs: List[Int]):(List[Int], List[Int])={
+  @tailrec
+  def inner(accum1: List[Int], accum2: List[Int], list: List[Int]):(List[Int], List[Int])={
+    list match{
+      case (Nil)=> (accum1, accum2);
+      case (h::t)=> if(h<0){
+        if(h%2!=0) inner(h::accum1, h::accum2, t)
+        else inner(h::accum1,accum2, t)
+      }else inner(accum1, accum2, t)
```

Comment:
struktura w tej okolicy do poprawy- strasznie ciężko się to czyta- chaotycznie połamane sekcje kodu